anesthetic/utils.py

```diff
@@ -89,6 +90,151 @@ def quantile(a, q, w=None, interpolation='linear'):
+    expect discretization errors on the scale of distances between samples.
+
+    https://github.com/Stefan-Heimersheim/fastCI#readme
+
+    Parameters
+    ----------
+    samples : array
+        Samples to compute the credibility interval of.
+    weights : array, default=np.ones_like(samples)
+        Weights corresponding to samples.
+    level : float, default=0.68
+        Credibility level (probability, <1).
+    method : str, default='iso-pdf'
+        Which definition of interval to use:
+
+        * ``'iso-pdf'``: Calculate iso probability density interval with the
+          same probability density at each end. Also known as
+          waterline-interval or highest average posterior density interval.
+          This is only accurate if the distribution is sufficiently unimodal.
+        * ``'lower-limit'``/``'upper-limit'``: Lower/upper limit. One-sided
+          limits for which ``level`` fraction of the (equally weighted) samples
+          lie above/below the limit.
+        * ``'equal-tailed'``: Equal-tailed interval with the same fraction of
+          (equally weighted) samples below and above the interval region.
+
+    return_covariance: bool, default=False
+        Return the covariance of the sampled limits, in addition to the mean
+    nsamples : int, default=12
+        Number of CDF samples to improve `mean` and `std` estimate.
+    verbose: bool, default=False
+        Print information and outputs.
+
+    Returns
+    -------
+    limit(s) : float, array, or tuple of floats or arrays
+        Returns the credibility interval boundari(es). By default
+        returns the mean over ``nsamples`` samples, which is either
+        two numbers (``method='iso-pdf'``/``'equal-tailed'``) or one number
+        (``method='lower-limit'``/``'upper-limit'``). If
+        ``return_covariance=True``, returns a tuple (mean(s), covariance)
+        where covariance is the covariance over the sampled limits.
+    """
+    if level >= 1:
+        raise ValueError('level must be <1, got {0:.2f}'.format(level))
+    if len(np.shape(samples)) != 1:
+        raise ValueError('Support only 1D arrays for samples')
+    if weights is not None and np.shape(samples) != np.shape(weights):
+        raise ValueError('Shape of samples and weights differs')
+
+    # Convert to numpy to unify indexing
+    samples = np.array(samples.copy())
+    if weights is None:
+        weights = np.ones(len(samples))
+    else:
+        weights = np.array(weights.copy())
+
+    # Convert samples to unit weight not the case
+    if not np.all(np.logical_or(weights == 0, weights == 1)):
+        # compress_weights with ncompress<=0 assures weights \in 0,1
+        # Note that this must be done, we cannot handle weights != 1
+        # see this discussion for details:
+        # https://github.com/williamjameshandley/anesthetic/pull/188#issuecomment-1274980982
+        weights = compress_weights(weights, ncompress=-1)
```

Comment:
We need `ncompress=-1` because the method does not work for weights other than 1. In practice this shouldn't matter because the (original) point of this method is to deal with huge numbers of samples.

Reply:
`samples.compress` will always return equally weighted samples (i.e. samples with all weights equal to 1), also for `ncompress=True` or for `ncompress=100000`. 

So are equal weights the only requirement? Or is there the additional requirement that there must not be repeated samples?

If only equal weights are needed, then `ncompress=True` will lead to a more conservative (and hence trustworthy...?) compression. 

If samples are not allowed to repeat, then indeed `ncompress=-1` is needed.

Reply:
With #285 we will need to change `ncompress=-1` to `ncompress='equal'`.